FirebasePhoneAuthUI/FUIPhoneEntryViewController.m

```diff
@@ -168,28 +168,36 @@ - (void)onNext:(NSString *)phoneNumber {
   [provider verifyPhoneNumber:phoneNumberWithCountryCode
                    UIDelegate:self
                    completion:^(NSString *_Nullable verificationID, NSError *_Nullable error) {
+    // Temporary fix to guarantee execution of the completion block on the main thread.
+    // TODO: Remove temporary workaround when the issue is fixed in FirebaseAuth.
+    dispatch_block_t completionBlock = ^() {
+      [self decrementActivity];
+      self.navigationItem.rightBarButtonItem.enabled = YES;
+
+      if (error) {
+        [_phoneNumberField becomeFirstResponder];
+
+        UIAlertController *alertController = [FUIPhoneAuth alertControllerForError:error
+                                                                     actionHandler:nil];
+        [self presentViewController:alertController animated:YES completion:nil];
+        
+        FUIPhoneAuth *delegate = [self.authUI providerWithID:FIRPhoneAuthProviderID];
+        [delegate callbackWithCredential:nil error:error result:nil];
+        return;
+      }
 
-    [self decrementActivity];
-    self.navigationItem.rightBarButtonItem.enabled = YES;
-
-    if (error) {
-      [_phoneNumberField becomeFirstResponder];
+      UIViewController *controller =
+          [[FUIPhoneVerificationViewController alloc] initWithAuthUI:self.authUI
+                                                      verificationID:verificationID
+                                                         phoneNumber:phoneNumberWithCountryCode];
 
-      UIAlertController *alertController = [FUIPhoneAuth alertControllerForError:error
-                                                                   actionHandler:nil];
-      [self presentViewController:alertController animated:YES completion:nil];
-      
-      FUIPhoneAuth *delegate = [self.authUI providerWithID:FIRPhoneAuthProviderID];
-      [delegate callbackWithCredential:nil error:error result:nil];
-      return;
+      [self pushViewController:controller];
+    };
+    if ([NSThread isMainThread]) {
+      completionBlock();
+    } else {
+      dispatch_sync(dispatch_get_main_queue(), completionBlock);
```

Comment:
I missed this earlier... I think we'd better to do dispatch_async instead of dispatch_sync, otherwise if the delegate calls FIRAuth methods there is a risk of deadlock.

Reply:
Fixed in 1eb511c2430546295fc5b6a03786511f71932a0a